src/execute.cpp

```diff
@@ -108,13 +114,16 @@ win32execute(const char* path,
   std::string args = Win32Util::argv_to_string(argv, sh);
   std::string full_path = Win32Util::add_exe_suffix(path);
   std::string tmp_file_path;
+  const char* lpAppName = full_path.c_str();
   if (args.length() > 8192) {
-    TemporaryFile tmp_file(path);
+    TemporaryFile tmp_file(FMT("{}/cmd_args", temp_dir));
```

Comment:
How about placing the temporary file in `%TEMP%` (with some appropriate fallback if that environment variable isn't set, I guess?) instead to avoid the need for a `temp_dir` parameter?

Reply:
I don't think it is good idea - using system temp directory when Ccache already has its own temp directory. It is weird externally visible behavior just because we didn't want to pass extra parameter.
I think we need to find good solution for execv() macro issue (ugly `saved_temp_dir` variable I added in `cache_compilation()`).

We could try to get rid of direct call to execv(). Maybe adding such version of execute:
`int execute_noreturn(const char* const* argv, Fd&& fd_out, Fd&& fd_err, const std::string& temp_dir);`
I would call this new function in `cache_compilation` (instead of execv()).
There is still `temp_dir` parameter, but now it is passed explicitly to called function, no surprising macro magic (I mean execv() macro in system.hpp).
This `temp_dir` parameter won't be used by Linux version of `execute_noreturn` but I think this is not a big deal.

Implementation of execute_noreturn() for Linux will be trivial - just a wrapper for execv (only `argv` parameter is used).
Implementation for Windows is also simple - a wrapper for win32_execute, similar to existing `execute`.

What do you think? Is it acceptable approach?

Reply:
IMHO this approach is a step toward better internal architecture of Ccache. I mean well architectured portability - instead of pretending on Windows that it is Linux (by macros like this execv()) there should be explicit "porting layer". `execute_noreturn` says what it does - it executes without return. It has different implementation on Windows and Linux. I'll try to implement it.

Reply:
Sure, steps towards improving internal architecture are welcome. The Windows port has been bolted on by several different contributors during the years. I have accepted the contributions in their current form since (a) I feel they have been unintrusive enough and (b) I know it's a large effort to restructure code that is/was inherently Unix-only (n.b.: not _Linux_-only) from the start, so I can't require more elaborate restructuring from contributors. I also often simply don't have time to do it myself. It's just how it is.

That said, I see nothing weird with using a predetermined temporary directory for temporary files. That's what most tools do all the time. The `temporary_dir` setting is actually legacy from how ccache worked before version 3.0 (released 2010). The setting could simply be removed, but I guess it still may have _some_ value since potentially large preprocessed output files are written there, so one might still want to be able to configure it. Maybe. I don't expect argument lists to be that large, though.